packages/backend-storage/src/action_to_resources_map.ts

```diff
@@ -0,0 +1,72 @@
+import { ResourceAccessAcceptor } from '@aws-amplify/plugin-types';
+import { StorageAction, StoragePrefix } from './types.js';
+
+/**
+ * Internal collaborating class for maintaining the relationship between an acceptor token and the access map
+ */
+export class AcceptorTokenAccessMap {
+  /**
+   * Maintains a mapping of actions for each token
+   */
+  private acceptorTokenAccessMap = new Map<AcceptorToken, AccessEntry>();
+
+  set = (
+    resourceAccessAcceptor: ResourceAccessAcceptor,
+    actions: StorageAction[],
+    s3Prefix: S3Prefix
+  ) => {
+    const acceptorToken = resourceAccessAcceptor.identifier;
+
+    if (!this.acceptorTokenAccessMap.has(acceptorToken)) {
+      this.acceptorTokenAccessMap.set(acceptorToken, {
+        actionMap: new S3PrefixActionMap(),
+        acceptor: resourceAccessAcceptor,
+      });
+    }
+    const actionMap = this.acceptorTokenAccessMap.get(acceptorToken)!.actionMap;
+    actions.forEach((action) => {
+      actionMap.set(action, s3Prefix);
+    });
+  };
+
+  getAccessList = () => {
+    const result: AccessEntry[] = [];
+    this.acceptorTokenAccessMap.forEach((value) => {
+      result.push(value);
+    });
+    return result as Readonly<Readonly<AccessEntry>[]>;
```

Comment:
if we use `Set` for `result`, is it possible to avoid `as`?

Reply:
The type assertion is to wrap everything in `Readonly`. We could do `as Readonly<Set<Readonly<AccessEntry>>>` instead but I don't think it really provides any benefit